File: parser.py
import re
import json

def _clean(value, fallback=""):
    if value is None:
        return fallback
    cleaned = value.strip()
    return cleaned if cleaned else fallback


def parse_markdown(content):
    result = {
        "metadata": {"title": "Untitled", "category": "General", "subcategory": None, "description": ""},
        "reading": [],
        "formulas": [],
        "mcqs": [],
        "notes": [],
        "todos": [],
    }

    lines = content.split('\n')
    current_section = "metadata"
    section_buffer = []
    metadata_done = False
    order_idx = 0
    in_code_block = False

    # Collect all h2-delimited sections
    sections = {} 
    current_h2 = None

    for line in lines:
        if line.startswith('```'):
            in_code_block = not in_code_block
        if line.startswith('# ') and not in_code_block:
            title = _clean(line[2:])
            if title:
                result["metadata"]["title"] = title
        elif line.startswith('## ') and not in_code_block:
            if current_h2 and section_buffer:
                sections[current_h2] = '\n'.join(section_buffer).strip()
            current_h2 = _clean(line[3:]).lower().replace(' ', '_')
            section_buffer = []
        else:
            section_buffer.append(line)

    if current_h2 and section_buffer:
        sections[current_h2] = '\n'.join(section_buffer).strip()

    # Extract metadata from ## Metadata section
    meta_text = sections.get('metadata', '')
    for line in meta_text.split('\n'):
        line = line.strip()
        lower = line.lower()
        if lower.startswith('- category:'):
            result["metadata"]["category"] = _clean(line.split(':', 1)[1], "General")
        elif lower.startswith('- subcategory:'):
            val = _clean(line.split(':', 1)[1])
            result["metadata"]["subcategory"] = val if val else None
        elif lower.startswith('- description:'):
            result["metadata"]["description"] = _clean(line.split(':', 1)[1])

    # Parse reading content (from ## Content or ## Reading)
    reading_key = None
    for key in ['content', 'reading']:
        if key in sections:
            reading_key = key
            break

    if reading_key:
        reading_text = sections[reading_key]
        # Split into h3 sections
        reading_sections = re.split(r'\n(?=### )', reading_text)
        for sec in reading_sections:
            sec = sec.strip()
            if not sec:
                continue
            title = ""
            content_lines = []
            for line in sec.split('\n'):
                if line.startswith('### '):
                    title = _clean(line[4:])
                else:
                    content_lines.append(line)
            body = '\n'.join(content_lines).strip()
            if body:
                result["reading"].append({
                    "title": title or "Section",
                    "content": body,
                    "order_idx": order_idx
                })
                order_idx += 1

    # Parse formulas from ## Formulas section
    formulas_text = sections.get('formulas', '')
    if formulas_text:
        for line in formulas_text.split('\n'):
            line = line.strip()
            if not line:
                continue
            m = re.match(r'\*\*(.+?):\*\*\s*(.+)', line)
            if m:
                title = _clean(m.group(1))
                content = _clean(m.group(2))
                if title and content:
                    result["formulas"].append({
                        "title": title,
                        "content": content,
                        "order_idx": order_idx
                    })
                    order_idx += 1

    # Parse MCQs from ## MCQs section
    mcqs_text = sections.get('mcqs', '')
    if mcqs_text:
        mcq_blocks = re.split(r'\n(?=\*\*Q\d+:)', mcqs_text)
        for block in mcq_blocks:
            block = block.strip()
            if not block:
                continue
            q_match = re.search(r'\*\*Q\d+:\*\*\s*(.+?)(?=\n(?:-\s+[A-D]\)))', block, re.DOTALL)
            if not q_match:
                continue
            question = _clean(q_match.group(1))

            options = []
            opt_matches = re.findall(r'-\s*([A-D])\)\s*(.+)', block)
            for key, text in opt_matches:
                options.append({"key": key.upper(), "text": _clean(text)})

            answer_m = re.search(r'\*\*Answer:\*\*\s*([A-D])', block)
            answer = _clean(answer_m.group(1)).upper() if answer_m else ""

            difficulty_m = re.search(r'\*\*Difficulty:\*\*\s*(.+)', block)
            difficulty = _clean(difficulty_m.group(1), 'medium').lower() if difficulty_m else 'medium'

            expl_m = re.search(r'\*\*Explanation:\*\*\s*(.+)', block, re.DOTALL)
            explanation = _clean(expl_m.group(1)) if expl_m else ""

            if options and answer:
                result["mcqs"].append({
                    "question": question,
                    "options": json.dumps(options),
                    "answer": answer,
                    "explanation": explanation,
                    "difficulty": difficulty,
                    "order_idx": order_idx
                })
                order_idx += 1

    # Parse notes from ## Notes section (bullet points)
    notes_text = sections.get('notes', '')
    if notes_text:
        for line in notes_text.split('\n'):
            line = line.strip()
            if line.startswith('- '):
                note = _clean(line[2:])
                if note:
                    result["notes"].append({"content": note, "order_idx": order_idx})
                    order_idx += 1

    # Parse todos from ## Todos section
    todos_text = sections.get('todos', '')
    if todos_text:
        for line in todos_text.split('\n'):
            line = line.strip()
            if line.startswith('- [ ]'):
                todo = _clean(line[5:])
                if todo:
                    result["todos"].append({"content": todo, "is_completed": 0, "order_idx": order_idx})
                    order_idx += 1
            elif line.startswith('- [x]'):
                todo = _clean(line[5:])
                if todo:
                    result["todos"].append({"content": todo, "is_completed": 1, "order_idx": order_idx})
                    order_idx += 1
            elif line.startswith('- '):
                todo = _clean(line[2:])
                if todo:
                    result["todos"].append({"content": todo, "is_completed": 0, "order_idx": order_idx})
                    order_idx += 1

    return result

def get_or_create_category(db, name, parent_name=None, user_id=None):
    name = _clean(name, "General")
    parent_id = None
    if isinstance(parent_name, int):
        parent_id = parent_name
    elif parent_name:
        parent_name = _clean(parent_name, None)
        if parent_name:
            parent = db.find_one('categories', name=parent_name, parent_id=None, user_id=user_id)
            if parent:
                parent_id = parent['id']
            else:
                parent_id = db.insert('categories', {'name': parent_name, 'parent_id': None, 'user_id': user_id})

    category = db.find_one('categories', name=name, parent_id=parent_id, user_id=user_id)
    if category:
        return category['id']
    else:
        return db.insert('categories', {'name': name, 'parent_id': parent_id, 'user_id': user_id})

File: src/backend/parser.js
function clean(value, fallback = '') {
  if (value == null) return fallback
  const s = String(value).trim()
  return s || fallback
}

function parseMarkdown(content) {
  const result = {
    metadata: { title: 'Untitled', category: 'General', subcategory: null, description: '' },
    reading: [],
    formulas: [],
    mcqs: [],
    notes: [],
    todos: [],
  }

  const lines = content.split('\n')
  let orderIdx = 0
  let inCodeBlock = false
  let currentH2 = null
  const sections = {}
  let sectionBuffer = []

  for (const line of lines) {
    if (line.startsWith('```')) {
      inCodeBlock = !inCodeBlock
      sectionBuffer.push(line)
      continue
    }
    if (line.startsWith('# ') && !inCodeBlock) {
      const title = clean(line.slice(2))
      if (title) result.metadata.title = title
      sectionBuffer.push(line)
      continue
    }
    if (line.startsWith('## ') && !inCodeBlock) {
      if (currentH2 && sectionBuffer.length > 0) {
        sections[currentH2] = sectionBuffer.join('\n').trim()
      }
      currentH2 = clean(line.slice(3)).toLowerCase().replace(/\s+/g, '_')
      sectionBuffer = []
    } else {
      sectionBuffer.push(line)
    }
  }
  if (currentH2 && sectionBuffer.length > 0) {
    sections[currentH2] = sectionBuffer.join('\n').trim()
  }

  const metaText = sections.metadata || ''
  for (const line of metaText.split('\n')) {
    const trimmed = line.trim()
    const lower = trimmed.toLowerCase()
    if (lower.startsWith('- category:')) {
      result.metadata.category = clean(trimmed.split(':')[1], 'General')
    } else if (lower.startsWith('- subcategory:')) {
      const val = clean(trimmed.split(':')[1])
      result.metadata.subcategory = val || null
    } else if (lower.startsWith('- description:')) {
      result.metadata.description = clean(trimmed.split(':')[1])
    }
  }

  let readingKey = null
  for (const key of ['content', 'reading']) {
    if (sections[key]) { readingKey = key; break }
  }

  if (readingKey) {
    const readingText = sections[readingKey]
    const readingSections = readingText.split(/\n(?=### )/)
    for (const sec of readingSections) {
      const trimmed = sec.trim()
      if (!trimmed) continue
      let title = ''
      const contentLines = []
      for (const line of trimmed.split('\n')) {
        if (line.startsWith('### ')) {
          title = clean(line.slice(4))
        } else {
          contentLines.push(line)
        }
      }
      const body = contentLines.join('\n').trim()
      if (body) {
        result.reading.push({ title: title || 'Section', content: body, order_idx: orderIdx })
        orderIdx++
      }
    }
  }

  const formulasText = sections.formulas || ''
  if (formulasText) {
    for (const line of formulasText.split('\n')) {
      const trimmed = line.trim()
      if (!trimmed) continue
      const m = trimmed.match(/\*\*(.+?):\*\*\s*(.+)/)
      if (m) {
        const title = clean(m[1])
        const content = clean(m[2])
        if (title && content) {
          result.formulas.push({ title, content, order_idx: orderIdx })
          orderIdx++
        }
      }
    }
  }

  const mcqsText = sections.mcqs || ''
  if (mcqsText) {
    const mcqBlocks = mcqsText.split(/\n(?=\*\*Q\d+:)/)
    for (const block of mcqBlocks) {
      const b = block.trim()
      if (!b) continue
      const qMatch = b.match(/\*\*Q\d+:\*\*\s*(.+?)(?=\n(?:-\s+[A-D]\)))/s)
      if (!qMatch) continue
      const question = clean(qMatch[1])

      const options = []
      const optMatches = b.matchAll(/-\s*([A-D])\)\s*(.+)/g)
      for (const match of optMatches) {
        options.push({ key: match[1].toUpperCase(), text: clean(match[2]) })
      }

      const answerM = b.match(/\*\*Answer:\*\*\s*([A-D])/)
      const answer = answerM ? clean(answerM[1]).toUpperCase() : ''

      const difficultyM = b.match(/\*\*Difficulty:\*\*\s*(.+)/)
      const difficulty = difficultyM ? clean(difficultyM[1], 'medium').toLowerCase() : 'medium'

      const explM = b.match(/\*\*Explanation:\*\*\s*(.+)/s)
      const explanation = explM ? clean(explM[1]) : ''

      if (options.length > 0 && answer) {
        result.mcqs.push({
          question,
          options: JSON.stringify(options),
          answer,
          explanation,
          difficulty,
          order_idx: orderIdx,
        })
        orderIdx++
      }
    }
  }

  const notesText = sections.notes || ''
  if (notesText) {
    for (const line of notesText.split('\n')) {
      const trimmed = line.trim()
      if (trimmed.startsWith('- ')) {
        const note = clean(trimmed.slice(2))
        if (note) {
          result.notes.push({ content: note, order_idx: orderIdx })
          orderIdx++
        }
      }
    }
  }

  const todosText = sections.todos || ''
  if (todosText) {
    for (const line of todosText.split('\n')) {
      const trimmed = line.trim()
      if (trimmed.startsWith('- [ ]')) {
        const todo = clean(trimmed.slice(5))
        if (todo) {
          result.todos.push({ content: todo, is_completed: 0, order_idx: orderIdx })
          orderIdx++
        }
      } else if (trimmed.startsWith('- [x]')) {
        const todo = clean(trimmed.slice(5))
        if (todo) {
          result.todos.push({ content: todo, is_completed: 1, order_idx: orderIdx })
          orderIdx++
        }
      } else if (trimmed.startsWith('- ')) {
        const todo = clean(trimmed.slice(2))
        if (todo) {
          result.todos.push({ content: todo, is_completed: 0, order_idx: orderIdx })
          orderIdx++
        }
      }
    }
  }

  return result
}

async function getOrCreateCategory(db, name, parentName = null, userId = null) {
  const catName = clean(name, 'General')
  let parentId = null

  if (typeof parentName === 'number') {
    parentId = parentName
  } else if (parentName) {
    const pName = clean(parentName, null)
    if (pName) {
      let parent = db.findOne('categories', { name: pName, parent_id: null, user_id: userId })
      if (parent) {
        parentId = parent.id
      } else {
        parentId = await db.insert('categories', { name: pName, parent_id: null, user_id: userId })
      }
    }
  }

  let category = db.findOne('categories', { name: catName, parent_id: parentId, user_id: userId })
  if (category) return category.id
  return await db.insert('categories', { name: catName, parent_id: parentId, user_id: userId })
}

module.exports = { parseMarkdown, getOrCreateCategory }
